wfe2/wfe.go

```diff
@@ -814,17 +784,19 @@ func (wfe *WebFrontEndImpl) NewAccount(
 	// are authoritative. This saves us from adding latency to each request.
 	// Goroutines spun out below will respect a context deadline set by the
 	// ratelimits package and cannot be prematurely canceled by the requester.
-	txns := wfe.newNewAccountLimitTransactions(ip)
-	go wfe.checkNewAccountLimits(ctx, txns)
+	refundLimits := func() {}
+	go func() {
+		refundLimits = wfe.checkNewAccountLimits(ctx, ip)
+	}()
 
 	var newRegistrationSuccessful bool
 	var errIsRateLimit bool
 	defer func() {
-		if !newRegistrationSuccessful && !errIsRateLimit {
+		if !newRegistrationSuccessful && !errIsRateLimit && refundLimits != nil {
```

Comment:
As I mentioned below, this is a data race. Also, `refundLimits` is never `nil` (since it's assigned to a noop func at creation).

Reply:
I should have removed the noop assignment. However, I'm fairly certain that `refundLimits` _can_ be made `nil` by `wfe.checkNewOrderLimits()`.

Here's a simple example of what I believe would happen here:
```go
package main

import "fmt"

func checkFunc(f func()) {
	if f == nil {
		fmt.Println("f is nil")
	} else {
		fmt.Println("f is not nil")
	}
}

func returnsNil() func() {
	return nil
}

func main() {
	var myFunc func() = func() {}
	checkFunc(myFunc)
	
	myFunc = returnsNil()
	checkFunc(myFunc)
}
```
```
f is not nil
f is nil
```
https://go.dev/play/p/88C-vEgm1UF